helm/resource_release.go

```diff
@@ -689,17 +694,113 @@ func resourceDiff(ctx context.Context, d *schema.ResourceDiff, meta interface{})
 	}
 	debug("%s Release validated", logID)
 
-	// Set desired version from the Chart metadata if available
-	if len(c.Metadata.Version) > 0 {
-		return d.SetNew("version", c.Metadata.Version)
+	// FIXME: make this a function so we can re-use it
+	// do dry-run upgrade to get diff
+	name := d.Get("name").(string)
+	namespace := d.Get("namespace").(string)
+
+	actionConfig, err := m.GetHelmConfiguration(namespace)
+	if err != nil {
+		return err
+	}
+
+	// check if release exists
+	_, err = getRelease(m, actionConfig, name)
```

Comment:
Unfortunately, this is calling getRelease so early in the process that it's sort of undoing the deferred provider initialization. What I mean by that is, it appears to try to contact the Kubernetes cluster before the credentials are available in the provider. And so it ends up having no provider configuration at all. [Here's my debug log](https://gist.githubusercontent.com/dak1n1/b92d6336f50a1d5bcc0afbd247895ab1/raw/084638ccaa462ea63defc687a0cfd162db01bbf6/gistfile1.txt).

These are the most relevant logs:
```
[DEBUG] getRelease for nginx-ingress-controller errored: timestamp=2021-03-20T12:47:16.506-0700
[DEBUG] Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable: timestamp=2021-03-20T12:47:16.506-0700
```

This would mean that users would no longer be able to install their cluster (GKE, EKS, AKS, etc) in the same apply as using the helm provider. However, when I run this same test from the `main` branch, it works fine.

If you want, you can run my test here:

https://github.com/hashicorp/terraform-provider-kubernetes/tree/master/kubernetes/test-infra/eks

You'll have to use terraform version 0.14.x because the EKS module doesn't work with 0.15 yet. Also be sure to unset KUBE_CONFIG_PATH and KUBE_CONFIG_PATHS, since that overrides the explicit provider configuration.

Reply:
It looks like it's erroring at plan time. Would it be possible to only use the new diff feature if there is an existing helm release in state, instead? That would at least allow the initial cluster build to succeed.

```
$ terraform0.14.7 plan

Error: error retrieving old release for a diff: Kubernetes cluster unreachable: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable

  on kubernetes-config/main.tf line 59, in resource "helm_release" "nginx_ingress":
  59: resource helm_release nginx_ingress {
```

Reply:
This is a good catch, thanks for testing this @dak1n1. I've added a check so that the custom diff only runs if the release already exists – can you test again?

Reply:
It works! 🎉